src/scenic/syntax/ast.py

```diff
@@ -1,20 +1,187 @@
 import ast
+from typing import Optional, Union
+
 
 class AST(ast.AST):
     "Scenic AST base class"
+
     def __init__(self, *args: any, **kwargs: any) -> None:
         super().__init__(*args, **kwargs)
 
+
 # special statements
 class EgoAssign(AST):
+    __match_args__ = ("value",)
+
     def __init__(self, value: any, *args: any, **kwargs: any) -> None:
         super().__init__(*args, **kwargs)
         self.value = value
         self._fields = ["value"]
 
+
+# Instance Creation
+
+
 class New(AST):
-    def __init__(self, className: str, specifiers: list, *args: any, **kwargs: any) -> None:
+    __match_args__ = ("className", "specifiers")
+
+    def __init__(
+        self, className: str, specifiers: list, *args: any, **kwargs: any
+    ) -> None:
         super().__init__(*args, **kwargs)
         self.className = className
         self.specifiers = specifiers
         self._fields = ["className", "specifiers"]
+
+
+# Specifiers
+
+
+class WithSpecifier(AST):
+    __match_args__ = ("prop", "value")
+
+    def __init__(self, prop: str, value: ast.AST, *args: any, **kwargs: any) -> None:
+        super().__init__(*args, **kwargs)
+        self.prop = prop
+        self.value = value
+        self._fields = ["prop", "value"]
+
+
+class AtSpecifier(AST):
+    __match_args__ = ("position",)
+
+    def __init__(self, position: ast.AST, *args: any, **kwargs: any) -> None:
+        super().__init__(*args, **kwargs)
+        self.position = position
+        self._fields = ["position"]
+
+
+class OffsetBySpecifier(AST):
+    __match_args__ = ("offset",)
+
+    def __init__(self, offset: ast.AST, *args: any, **kwargs: any) -> None:
+        super().__init__(*args, **kwargs)
+        self.offset = offset
+        self._fields = ["offset"]
+
+
+class OffsetAlongSpecifier(AST):
+    __match_args__ = ("direction", "offset")
+
+    def __init__(
+        self, direction: ast.AST, offset: ast.AST, *args: any, **kwargs: any
+    ) -> None:
+        super().__init__(*args, **kwargs)
+        self.direction = direction
+        self.offset = offset
+        self._fields = ["direction", "offset"]
+
+
+class PositionSpecifier(AST):
+    __match_args__ = ("direction", "position", "distance")
+
+    def __init__(
+        self,
+        direction: Union["LeftOf", "RightOf", "AheadOf", "Behind"],
+        position: ast.AST,
+        distance: Optional[ast.AST],
+        *args: any,
+        **kwargs: any
+    ) -> None:
+        super().__init__(*args, **kwargs)
+        self.direction = direction
+        self.position = position
+        self.distance = distance
+        self._fields = ["direction", "position", "distance"]
+
+
+class LeftOf(AST):
```

Comment:
On second thought, the reason CPython uses AST nodes might be that their AST nodes are auto-generated by ASDL definition and it does not support string literals.

https://github.com/python/cpython/blob/e69306f08b9be84ccdd0a1c6601ec229c4e5b377/Parser/Python.asdl
https://www.cs.princeton.edu/research/techreps/TR-554-97

Another difference I can think of is how NodeVisitor works on these classes.

```python
import ast

class V(ast.NodeVisitor):
    def visit_Not(self, node):
        print("found", ast.dump(node))

V().visit(ast.parse("not x"))
```

For example, the code above would print "found Not()". We can define `visit_Not` because Not is an AST node, and if it was some kind of ENUM representation (string literal, number, etc.), we would have to check `op` inside `visit_UnaryOp`.

I'd appreciate hearing your opinion.

Reply:
Ah, OK, if it's more uniform we can just keep them as AST nodes.